pkg/controllers/ragengine_controller.go

```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
+			return nil, updateErr
+		}
+		return nil, err
+	}
+
+	// get the node object from the nodeClaim status nodeName.
+	return resources.GetNode(ctx, newNodeClaim.Status.NodeName, c.Client)
+}
+
+// ensureNodePlugins ensures node plugins are installed.
+func (c *RAGEngineReconciler) ensureNodePlugins(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine, nodeObj *corev1.Node) error {
+	timeClock := clock.RealClock{}
+	tick := timeClock.NewTicker(nodePluginInstallTimeout)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-tick.C():
+			return fmt.Errorf("node plugin installation timed out. node %s is not ready", nodeObj.Name)
+		default:
+			//Nvidia Plugin
+			if found := resources.CheckNvidiaPlugin(ctx, nodeObj); !found {
+				if err := resources.UpdateNodeWithLabel(ctx, nodeObj.Name, resources.LabelKeyNvidia, resources.LabelValueNvidia, c.Client); err != nil {
+					if apierrors.IsNotFound(err) {
+						klog.ErrorS(err, "nvidia plugin cannot be installed, node not found", "node", nodeObj.Name)
+						if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+							if updateErr := c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeNodeClaimStatus, metav1.ConditionFalse,
+								"checkNodeClaimStatusFailed", err.Error()); updateErr != nil {
+								klog.ErrorS(updateErr, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+								return updateErr
+							}
+						} else {
+							if updateErr := c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeMachineStatus, metav1.ConditionFalse,
+								"checkMachineStatusFailed", err.Error()); updateErr != nil {
+								klog.ErrorS(updateErr, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+								return updateErr
+							}
+						}
+						return err
+					}
+				}
+				time.Sleep(1 * time.Second)
+			}
+			return nil
+		}
+	}
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (c *RAGEngineReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	c.Recorder = mgr.GetEventRecorderFor("RAGEngine")
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{},
+		"spec.nodeName", func(rawObj client.Object) []string {
+			pod := rawObj.(*corev1.Pod)
+			return []string{pod.Spec.NodeName}
+		}); err != nil {
+		return err
+	}
 	builder := ctrl.NewControllerManagedBy(mgr).
 		For(&kaitov1alpha1.RAGEngine{}).
+		Watches(&v1alpha5.Machine{}, c.watchMachines()).
 		WithOptions(controller.Options{MaxConcurrentReconciles: 5})
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		builder.
+			Watches(&v1beta1.NodeClaim{}, c.watchNodeClaims()) // watches for nodeClaim with labels indicating ragengine name.
+	}
 	return builder.Complete(c)
 }
+
+// watches for machine with labels indicating RAGEngine name.
+func (c *RAGEngineReconciler) watchMachines() handler.EventHandler {
```

Comment:
same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L890